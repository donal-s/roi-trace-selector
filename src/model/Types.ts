export const SCANSTATUS_SELECTED = "y";
export const SCANSTATUS_UNSELECTED = "n";
export const SCANSTATUS_UNSCANNED = "?";

export type ScanStatus =
  | typeof SCANSTATUS_SELECTED
  | typeof SCANSTATUS_UNSELECTED
  | typeof SCANSTATUS_UNSCANNED;

export type SelectedItemCounts = {
  selectedCount: number;
  unselectedCount: number;
  unscannedCount: number;
};

export type ChartAlignment = {
  enableYMaxAlignment: boolean;
  alignToYMax: boolean;
  yMaxValue: number;
  yMaxFrame: number;
  enableYMinAlignment: boolean;
  alignToYMin: boolean;
  yMinValue: number;
  yMinFrame: number;
};

export type DataFile = {
  csvData: string;
  channel1Filename: string | null;
};

export const AXIS_H = "h";
export const AXIS_V = "v";
export type Axis = typeof AXIS_H | typeof AXIS_V;

export type Annotation = {
  name: string;
  axis: Axis;
  value: number;
};

export type EditAnnotation = {
  index: number;
  annotation: Annotation;
};
