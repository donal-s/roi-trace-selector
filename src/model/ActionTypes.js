export const LOAD_DATA = "loadData";
export const SET_FULLSCREEN_MODE = "setFullscreenMode";
export const SET_CURRENT_INDEX = "setCurrentIndex";
export const SET_CURRENT_NEXT = "setCurrentNext";
export const SET_CURRENT_PREVIOUS = "setCurrentPrevious";
export const SET_CURRENT_NEXT_UNSCANNED = "setCurrentNextUnscanned";
export const SET_CURRENT_SCANSTATUS = "setCurrentScanStatus";
export const TOGGLE_CURRENT_ITEM_SELECTED = "toggleCurrentItemSelected";
export const SELECT_ALL_ITEMS = "selectAllItems";
export const UPDATE_CHART_ALIGNMENT = "updateChartAlignment";

// For testing
export const RESET_STATE = "resetState";
