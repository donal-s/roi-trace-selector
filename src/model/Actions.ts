import { createAction } from "@reduxjs/toolkit";
import {
  ChartAlignment,
  ScanStatus,
  ChannelData,
  SCANSTATUS_SELECTED,
  SCANSTATUS_UNSCANNED,
  SCANSTATUS_UNSELECTED,
  Annotation,
  EditAnnotation,
  Channel,
} from "./Types";

function withPayloadType<T>() {
  return (t: T) => ({ payload: t });
}

export const fullscreenModeAction = createAction(
  "setFullscreenMode",
  withPayloadType<boolean>()
);
export const setCurrentIndexAction = createAction(
  "setCurrentIndex",
  withPayloadType<number>()
);
export const setCurrentNextAction = createAction("setCurrentNext");
export const setCurrentPreviousAction = createAction("setCurrentPrevious");
export const setCurrentNextUnscannedAction = createAction(
  "setCurrentNextUnscanned"
);
export const toggleCurrentItemSelectedAction = createAction(
  "toggleCurrentItemSelected"
);
export const selectAllItemsAction = createAction("selectAllItems");
export const updateChartAlignmentAction = createAction(
  "updateChartAlignment",
  withPayloadType<ChartAlignment>()
);
export const loadChannelAction = createAction(
  "loadChannel",
  withPayloadType<ChannelData>()
);
export const closeChannelAction = createAction(
  "closeChannel",
  withPayloadType<Channel>()
);
export const resetStateAction = createAction("resetState");

export const setCurrentScanStatusAction = createAction(
  "setCurrentScanStatus",
  withPayloadType<ScanStatus>()
);
export const setCurrentSelectedAction = () =>
  setCurrentScanStatusAction(SCANSTATUS_SELECTED);
export const setCurrentUnselectedAction = () =>
  setCurrentScanStatusAction(SCANSTATUS_UNSELECTED);
export const setCurrentUnscannedAction = () =>
  setCurrentScanStatusAction(SCANSTATUS_UNSCANNED);

export const updateAnnotationsAction = createAction(
  "updateAnnotations",
  withPayloadType<Annotation[]>()
);

export const updateEditAnnotationAction = createAction(
  "updateEditAnnotation",
  withPayloadType<EditAnnotation | undefined>()
);
