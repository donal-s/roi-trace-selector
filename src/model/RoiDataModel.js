import { createStore, applyMiddleware } from "redux";
import thunk from "redux-thunk";
import {
  SET_FULLSCREEN_MODE,
  SET_CURRENT_INDEX,
  SET_CURRENT_NEXT,
  SET_CURRENT_PREVIOUS,
  SET_CURRENT_NEXT_UNSCANNED,
  SET_CURRENT_SCANSTATUS,
  TOGGLE_CURRENT_ITEM_SELECTED,
  SELECT_ALL_ITEMS,
  UPDATE_CHART_ALIGNMENT,
  RESET_STATE,
  LOAD_DATA,
} from "./ActionTypes.js";
import {
  SCANSTATUS_SELECTED,
  SCANSTATUS_UNSELECTED,
  SCANSTATUS_UNSCANNED,
} from "./ScanStatus.js";
import { fileHandlingReducer } from "./CsvHandling.js";

const initialState = {
  channel1Filename: null,
  items: [],
  scanStatus: [],
  currentIndex: -1,
  chartFrameLabels: [],
  chartData: [],
  originalTraceData: [],
  showSingleTrace: false,
};

// Exported for unit tests only
export function roiDataReducer(state = initialState, action) {
  switch (action.type) {
    case SET_FULLSCREEN_MODE:
      return setFullscreenMode(state, action);

    case SET_CURRENT_INDEX:
      return setCurrentIndex(state, action);

    case SET_CURRENT_NEXT:
      return setCurrentNext(state, action);

    case SET_CURRENT_PREVIOUS:
      return setCurrentPrevious(state, action);

    case SET_CURRENT_NEXT_UNSCANNED:
      return setCurrentNextUnscanned(state, action);

    case SET_CURRENT_SCANSTATUS:
      return setCurrentScanStatus(state, action);

    case TOGGLE_CURRENT_ITEM_SELECTED:
      return toggleCurrentItemSelected(state, action);

    case SELECT_ALL_ITEMS:
      return selectAllItems(state, action);

    case UPDATE_CHART_ALIGNMENT:
      return updateChartAlignment(state, action);

    case RESET_STATE:
      return initialState;

    default:
      // Skip logging this one - CsvHandling will pick it up
      if (action.type !== LOAD_DATA) {
        console.log("Unknown action: " + safeJson(action));
      }
      return state;
  }
}

function setFullscreenMode(state, action) {
  return {
    ...state,
    showSingleTrace: action.enable,
  };
}

function setCurrentIndex(state, action) {
  checkIndex(state, action.index);

  return {
    ...state,
    currentIndex: action.index,
  };
}

function setCurrentNext(state, action) {
  if (state.items.length > 0 && state.currentIndex < state.items.length - 1) {
    return {
      ...state,
      currentIndex: state.currentIndex + 1,
    };
  }
  return state;
}

function setCurrentPrevious(state, action) {
  if (state.items.length > 0 && state.currentIndex > 0) {
    return {
      ...state,
      currentIndex: state.currentIndex - 1,
    };
  }
  return state;
}

function setCurrentNextUnscanned(state, action) {
  var itemCount = state.items.length;
  if (itemCount > 0) {
    for (var i = state.currentIndex + 1; i < itemCount; i++) {
      if (state.scanStatus[i] === SCANSTATUS_UNSCANNED) {
        return {
          ...state,
          currentIndex: i,
        };
      }
    }
    // Wrap around
    for (i = 0; i < state.currentIndex; i++) {
      if (state.scanStatus[i] === SCANSTATUS_UNSCANNED) {
        return {
          ...state,
          currentIndex: i,
        };
      }
    }
  }
  return state;
}

function setCurrentScanStatus(state, action) {
  if (state.currentIndex !== -1) {
    var scanStatus = [...state.scanStatus];
    scanStatus[state.currentIndex] = action.scanStatus;
    return {
      ...state,
      scanStatus: scanStatus,
    };
  }

  return state;
}

function toggleCurrentItemSelected(state, action) {
  const index = state.currentIndex;
  if (index !== -1) {
    checkIndex(state, index);
    var scanStatus = [...state.scanStatus];
    if (scanStatus[index] === SCANSTATUS_SELECTED) {
      scanStatus[index] = SCANSTATUS_UNSELECTED;
    } else if (scanStatus[index] === SCANSTATUS_UNSELECTED) {
      scanStatus[index] = SCANSTATUS_UNSCANNED;
    } else {
      scanStatus[index] = SCANSTATUS_SELECTED;
    }
    return {
      ...state,
      scanStatus: scanStatus,
    };
  }

  return state;
}

function selectAllItems(state, action) {
  const itemCounts = getSelectedItemCounts(state);
  const itemStatus = getSelectAllStatus(itemCounts);
  const itemCount = state.scanStatus.length;
  if (itemCount === 0) {
    return state;
  }
  var newScanStatus = Array(itemCount);
  newScanStatus.fill(itemStatus);

  return {
    ...state,
    scanStatus: newScanStatus,
  };
}

function updateChartAlignment(
  state,
  {
    enableYMaxAlignment,
    alignToYMax,
    yMaxValue,
    yMaxFrame,
    enableYMinAlignment,
    alignToYMin,
    yMinValue,
    yMinFrame,
  }
) {
  if (
    (enableYMaxAlignment &&
      !alignToYMax &&
      (yMaxFrame < 1 || yMaxFrame > getFrameCount(state))) ||
    (enableYMinAlignment &&
      !alignToYMin &&
      (yMinFrame < 1 || yMinFrame > getFrameCount(state)))
  ) {
    throw new Error("Invalid frame index: " + yMinFrame + ", " + yMaxFrame);
  }

  var roiCount = getItemCount(state);
  var frameCount = getFrameCount(state);

  var newChartData = [];

  for (var roiIndex = 0; roiIndex < roiCount; roiIndex++) {
    var inputRoi = state.originalTraceData[roiIndex];

    var outputRoi = [...state.chartData[roiIndex]];

    if (enableYMaxAlignment) {
      var rawYMaxValue;
      if (alignToYMax) {
        rawYMaxValue = state.originalTraceData[roiIndex][0];
        for (var frameIndex = 1; frameIndex < frameCount; frameIndex++) {
          rawYMaxValue = Math.max(rawYMaxValue, inputRoi[frameIndex]);
        }
      } else {
        rawYMaxValue = inputRoi[yMaxFrame - 1];
      }

      if (enableYMinAlignment) {
        var yScale = 1;
        var rawYMinValue;
        if (alignToYMin) {
          rawYMinValue = state.originalTraceData[roiIndex][0];
          for (frameIndex = 1; frameIndex < frameCount; frameIndex++) {
            rawYMinValue = Math.min(rawYMinValue, inputRoi[frameIndex]);
          }
        } else {
          rawYMinValue = state.originalTraceData[roiIndex][yMinFrame - 1];
        }
        if (rawYMaxValue === rawYMinValue) {
          yScale = 1;
        } else {
          yScale = (yMaxValue - yMinValue) / (rawYMaxValue - rawYMinValue);
        }

        for (frameIndex = 0; frameIndex < frameCount; frameIndex++) {
          outputRoi[frameIndex] =
            (inputRoi[frameIndex] - rawYMaxValue) * yScale + +yMaxValue;
        }
      } else {
        for (frameIndex = 0; frameIndex < frameCount; frameIndex++) {
          outputRoi[frameIndex] =
            inputRoi[frameIndex] - rawYMaxValue + +yMaxValue;
        }
      }
    } else {
      outputRoi = [...inputRoi];
    }
    newChartData.push(outputRoi);
  }

  return {
    ...state,
    chartData: newChartData,
  };
}

// Exported for unit tests
export function appReducer(state, action) {
  return fileHandlingReducer(roiDataReducer(state, action), action);
}

export default createStore(appReducer, applyMiddleware(thunk));

export const isChannel1Loaded = ({ items }) => items.length > 0;

function getSelectAllStatus([selectedCount, unselectedCount, unscannedCount]) {
  if (selectedCount === 0 && unselectedCount === 0) {
    return SCANSTATUS_SELECTED;
  } else if (unscannedCount === 0 && unselectedCount === 0) {
    return SCANSTATUS_UNSELECTED;
  } else {
    // everything else
    return SCANSTATUS_UNSCANNED;
  }
}

export function getSelectAllActionName([
  selectedCount,
  unselectedCount,
  unscannedCount,
]) {
  if (selectedCount === 0 && unselectedCount === 0) {
    return "Select All";
  } else if (unscannedCount === 0 && unselectedCount === 0) {
    return "Unselect All";
  } else {
    // everything else
    return "Clear All";
  }
}

export function getSelectedItemCounts(model) {
  var result = [0, 0, 0];
  model.scanStatus.forEach((item) => {
    if (item === SCANSTATUS_SELECTED) {
      result[0] = result[0] + 1;
    } else if (item === SCANSTATUS_UNSELECTED) {
      result[1] = result[1] + 1;
    } else {
      result[2] = result[2] + 1;
    }
  });
  return result;
}

export function isItemSelected(scanStatus, index) {
  checkIndex({ scanStatus }, index);
  return scanStatus[index] === SCANSTATUS_SELECTED;
}

function checkIndex({ scanStatus }, index) {
  if (index < 0 || index >= scanStatus.length) {
    throw new Error("ROI index not valid: " + index);
  }
}

export function isItemUnselected(scanStatus, index) {
  checkIndex({ scanStatus }, index);
  return scanStatus[index] === SCANSTATUS_UNSELECTED;
}

export function isCurrentSelected({ currentIndex, scanStatus }) {
  return (
    currentIndex !== -1 && scanStatus[currentIndex] === SCANSTATUS_SELECTED
  );
}

export function isCurrentUnselected({ currentIndex, scanStatus }) {
  return (
    currentIndex !== -1 && scanStatus[currentIndex] === SCANSTATUS_UNSELECTED
  );
}

export function isCurrentUnscanned({ currentIndex, scanStatus }) {
  return (
    currentIndex !== -1 && scanStatus[currentIndex] === SCANSTATUS_UNSCANNED
  );
}

export function setCurrentSelected() {
  return {
    type: SET_CURRENT_SCANSTATUS,
    scanStatus: SCANSTATUS_SELECTED,
  };
}

export function setCurrentUnselected() {
  return {
    type: SET_CURRENT_SCANSTATUS,
    scanStatus: SCANSTATUS_UNSELECTED,
  };
}

export function setCurrentUnscanned() {
  return {
    type: SET_CURRENT_SCANSTATUS,
    scanStatus: SCANSTATUS_UNSCANNED,
  };
}

function getItemCount({ items }) {
  return items.length;
}

export function getFrameCount({ chartFrameLabels }) {
  return chartFrameLabels.length;
}
/*
const getCircularReplacer = () => {
  const seen = new WeakSet();
  return (key, value) => {
    if (typeof value === "object" && value !== null) {
      if (seen.has(value)) {
        return;
      }
      seen.add(value);
    }
    return value;
  };
};*/

function safeJson(value) {
  //return JSON.stringify(value, getCircularReplacer());
  return JSON.stringify(value);
}
