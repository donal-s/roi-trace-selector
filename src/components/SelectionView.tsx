import React from "react";
import {
  getItemCount,
  useAppDispatch,
  useAppSelector,
} from "../model/RoiDataModel";
import { setSelectionAction } from "../model/Actions";
import {
  CHANNEL_1,
  Selection,
  SelectionManual,
  SelectionType,
  SELECTION_MANUAL,
  SELECTION_MINIMUM_STDEV_BY_TRACE_COUNT,
  SELECTION_PERCENT_CHANGE,
  SELECTION_STDEV,
} from "../model/Types";

const DEFAULT_SELECTION: SelectionManual = {
  type: SELECTION_MANUAL,
};

export default function SelectionView() {
  const dispatch = useAppDispatch();
  const currentChannel = useAppSelector((state) => state.currentChannel);
  const chartFrameLabels = useAppSelector((state) => state.chartFrameLabels);
  const channelSelection = useAppSelector((state) =>
    currentChannel === CHANNEL_1
      ? state.channel1Dataset?.selection
      : state.channel2Dataset?.selection
  );

  const selection: Selection = channelSelection || DEFAULT_SELECTION;

  const datasetFrameCount = useAppSelector(
    (state) => state.chartFrameLabels.length || 1
  );

  const datasetItemCount = useAppSelector(getItemCount);

  type CreateInputParams = {
    key: string;
    title: string;
    onBlur?: any;
    min?: number;
    max?: number;
    step?: number;
  };

  function createNumberInput({
    key,
    title,
    step,
    ...otherInputAttributes
  }: CreateInputParams) {
    return (
      <>
        <label htmlFor={key}>{title}</label>
        <input
          type="number"
          id={key}
          value={(selection as any)[key]}
          onChange={(event) => {
            if (event.target.value !== "") {
              dispatch(
                setSelectionAction({
                  ...selection,
                  [key]: Number(event.target.value),
                })
              );
            }
          }}
          step={step || "1"}
          {...otherInputAttributes}
        />
      </>
    );
  }

  function createPercentageChangeInput({
    key,
    title,
    step,
    ...otherInputAttributes
  }: CreateInputParams) {
    return (
      <>
        <label htmlFor={key}>{title}</label>
        <input
          type="number"
          id={key}
          value={Math.round((selection as any)[key] * 100)}
          onChange={(event) => {
            if (event.target.value !== "") {
              dispatch(
                setSelectionAction({
                  ...selection,
                  [key]: Number(event.target.value) / 100,
                })
              );
            }
          }}
          step={step || "1"}
          {...otherInputAttributes}
        />
      </>
    );
  }

  function createFrameNumberInput({ key, title }: CreateInputParams) {
    return (
      <>
        <label htmlFor={key}>{title}</label>
        <input
          type="number"
          id={key}
          value={chartFrameLabels[(selection as any)[key]]}
          onChange={(event) => {
            if (event.target.value !== "") {
              dispatch(
                setSelectionAction({
                  ...selection,
                  [key]: Number(event.target.value) - 1,
                })
              );
            }
          }}
          min={chartFrameLabels[0]}
          max={chartFrameLabels[datasetFrameCount - 1]}
        />
      </>
    );
  }

  function updateSelectionType(value: SelectionType) {
    if (value === selection.type) {
      return;
    }
    let newSelection: Selection;
    switch (value) {
      case SELECTION_PERCENT_CHANGE:
        newSelection = {
          type: value,
          startFrame: 0,
          endFrame: datasetFrameCount - 1,
          percentChange: 0.1,
        };
        break;
      case SELECTION_STDEV:
        newSelection = {
          type: value,
          startBaselineFrame: 0,
          endBaselineFrame: datasetFrameCount - 1,
          startDetectionFrame: 0,
          endDetectionFrame: datasetFrameCount - 1,
          stdevMultiple: 1,
        };
        break;
      case SELECTION_MINIMUM_STDEV_BY_TRACE_COUNT:
        newSelection = {
          type: value,
          selectedTraceCount: datasetItemCount,
          selectedStdev: 1,
        };
        break;
      default:
        newSelection = DEFAULT_SELECTION;
    }
    dispatch(setSelectionAction(newSelection));
  }

  return (
    <div id="selectionPanel">
      <select
        id="selectionMode"
        value={selection.type}
        onChange={(event) =>
          updateSelectionType(event.target.value as SelectionType)
        }
      >
        <option value={SELECTION_MANUAL}>Manual</option>
        <option value={SELECTION_PERCENT_CHANGE}>% Change</option>
        <option value={SELECTION_STDEV}>STDEV</option>
        <option value={SELECTION_MINIMUM_STDEV_BY_TRACE_COUNT}>
          Remove Outliers
        </option>
      </select>
      <div className="optionsPanel">
        {selection.type === SELECTION_PERCENT_CHANGE && (
          <>
            {createFrameNumberInput({
              key: "startFrame",
              title: "Start frame",
            })}
            {createFrameNumberInput({
              key: "endFrame",
              title: "End frame",
            })}
            {createPercentageChangeInput({
              key: "percentChange",
              title: "% of total change",
              min: 1,
              max: 100,
            })}
          </>
        )}

        {selection.type === SELECTION_STDEV && (
          <>
            {createFrameNumberInput({
              key: "startBaselineFrame",
              title: "Start baseline frame",
            })}
            {createFrameNumberInput({
              key: "endBaselineFrame",
              title: "End baseline frame",
            })}
            {createFrameNumberInput({
              key: "startDetectionFrame",
              title: "Start detection frame",
            })}
            {createFrameNumberInput({
              key: "endDetectionFrame",
              title: "End detection frame",
            })}
            {createNumberInput({
              key: "stdevMultiple",
              title: "x STDEV minimum",
              min: 0.1,
              max: 10,
              step: 0.1,
            })}
          </>
        )}

        {selection.type === SELECTION_MINIMUM_STDEV_BY_TRACE_COUNT && (
          <>
            {createNumberInput({
              key: "selectedTraceCount",
              title: "Selected trace count",
              min: 2,
              max: datasetItemCount,
            })}
            <label htmlFor="selectedStdev">STDEV of selected</label>
            <div id="selectedStdev">{selection.selectedStdev.toFixed(2)}</div>
          </>
        )}
      </div>
    </div>
  );
}
