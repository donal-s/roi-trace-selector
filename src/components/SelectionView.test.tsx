/* eslint jest/expect-expect: ["error", { "assertFunctionNames": ["expect", "checkPanel"] }] */

import React from "react";
import SelectionView from "./SelectionView";
import { RoiDataModelState } from "../model/RoiDataModel";
import {
  renderWithProvider,
  EMPTY_STATE,
  DUAL_CHANNEL_LOADED_STATE,
} from "../TestUtils";
import { screen, waitFor } from "@testing-library/react";
import { CHANNEL_2 } from "../model/Types";

describe("component SelectionView", () => {
  it("render selection choices - empty state", () => {
    renderComponent();

    // Manual selection
    expect(selectionSelect()).toHaveDisplayValue("Manual");
  });

  it("render selection choices - first channel", async () => {
    const { container, store, user } = renderComponent(
      DUAL_CHANNEL_LOADED_STATE,
    );

    // No selection = default to Manual selection
    expect(selectionSelect()).toHaveDisplayValue("Manual");
    expect(container).toMatchSnapshot();

    // Percent change selection
    await user.selectOptions(selectionSelect(), "% Change");

    await waitFor(() =>
      expect(store.getState().channel1Dataset?.selection).toEqual({
        endFrame: 2,
        percentChange: 0.1,
        startFrame: 0,
        type: "percentChange",
      }),
    );
    expect(screen.getByLabelText("Start frame")).toHaveValue(1);
    expect(screen.getByLabelText("End frame")).toHaveValue(3);
    expect(screen.getByLabelText("% of total change")).toHaveValue(10);

    expect(container).toMatchSnapshot();

    // STDEV selection
    await user.selectOptions(selectionSelect(), "STDEV");

    await waitFor(() =>
      expect(store.getState().channel1Dataset?.selection).toEqual({
        endBaselineFrame: 2,
        endDetectionFrame: 4,
        startBaselineFrame: 0,
        startDetectionFrame: 3,
        stdevMultiple: 1,
        type: "stdev",
      }),
    );

    expect(screen.getByLabelText("Start baseline frame")).toHaveValue(1);
    expect(screen.getByLabelText("End baseline frame")).toHaveValue(3);
    expect(screen.getByLabelText("Start detection frame")).toHaveValue(4);
    expect(screen.getByLabelText("End detection frame")).toHaveValue(5);
    expect(screen.getByLabelText("x STDEV minimum")).toHaveValue(1);

    expect(container).toMatchSnapshot();

    // Remove outliers
    await user.selectOptions(selectionSelect(), "Remove Outliers");

    await waitFor(() =>
      expect(store.getState().channel1Dataset?.selection).toEqual({
        selectedStdev: expect.closeTo(2.49),
        selectedTraceCount: 4,
        type: "minimumStdevByTraceCount",
      }),
    );

    expect(screen.getByLabelText("Selected trace count")).toHaveValue(4);
    expect(screen.getByLabelText("STDEV of selected")).toHaveTextContent(
      "2.49",
    );

    expect(container).toMatchSnapshot();
  });

  it("render selection choices - second channel", async () => {
    const { container, store, user } = renderComponent({
      ...DUAL_CHANNEL_LOADED_STATE,
      currentChannel: CHANNEL_2,
    });

    // No selection = default to Manual selection
    expect(selectionSelect()).toHaveDisplayValue("Manual");
    expect(container).toMatchSnapshot();

    // Percent change selection
    await user.selectOptions(selectionSelect(), "% Change");

    await waitFor(() =>
      expect(store.getState().channel2Dataset?.selection).toEqual({
        endFrame: 2,
        percentChange: 0.1,
        startFrame: 0,
        type: "percentChange",
      }),
    );
    expect(screen.getByLabelText("Start frame")).toHaveValue(1);
    expect(screen.getByLabelText("End frame")).toHaveValue(3);
    expect(screen.getByLabelText("% of total change")).toHaveValue(10);

    expect(container).toMatchSnapshot();

    // STDEV selection
    await user.selectOptions(selectionSelect(), "STDEV");

    await waitFor(() =>
      expect(store.getState().channel2Dataset?.selection).toEqual({
        endBaselineFrame: 2,
        endDetectionFrame: 4,
        startBaselineFrame: 0,
        startDetectionFrame: 3,
        stdevMultiple: 1,
        type: "stdev",
      }),
    );

    expect(screen.getByLabelText("Start baseline frame")).toHaveValue(1);
    expect(screen.getByLabelText("End baseline frame")).toHaveValue(3);
    expect(screen.getByLabelText("Start detection frame")).toHaveValue(4);
    expect(screen.getByLabelText("End detection frame")).toHaveValue(5);
    expect(screen.getByLabelText("x STDEV minimum")).toHaveValue(1);

    expect(container).toMatchSnapshot();

    // Remove outliers
    await user.selectOptions(selectionSelect(), "Remove Outliers");

    await waitFor(() =>
      expect(store.getState().channel2Dataset?.selection).toEqual({
        selectedStdev: expect.closeTo(2.49),
        selectedTraceCount: 4,
        type: "minimumStdevByTraceCount",
      }),
    );

    expect(screen.getByLabelText("Selected trace count")).toHaveValue(4);
    expect(screen.getByLabelText("STDEV of selected")).toHaveTextContent(
      "2.49",
    );

    expect(container).toMatchSnapshot();
  });

  const selectionSelect = (): HTMLElement => document.querySelector("select")!;

  function renderComponent(preloadedState: RoiDataModelState = EMPTY_STATE) {
    return renderWithProvider(<SelectionView />, { preloadedState });
  }
});
