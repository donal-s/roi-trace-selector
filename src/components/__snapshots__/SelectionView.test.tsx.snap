// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component SelectionView render selection choices - first channel 1`] = `
<div>
  <div
    id="selectionPanel"
  >
    <select
      id="selectionMode"
    >
      <option
        value="manual"
      >
        Manual
      </option>
      <option
        value="percentChange"
      >
        % Change
      </option>
      <option
        value="stdev"
      >
        STDEV
      </option>
      <option
        value="minimumStdevByTraceCount"
      >
        Remove Outliers
      </option>
    </select>
    <div
      class="optionsPanel"
    />
  </div>
</div>
`;

exports[`component SelectionView render selection choices - first channel 2`] = `
<div>
  <div
    id="selectionPanel"
  >
    <select
      id="selectionMode"
    >
      <option
        value="manual"
      >
        Manual
      </option>
      <option
        value="percentChange"
      >
        % Change
      </option>
      <option
        value="stdev"
      >
        STDEV
      </option>
      <option
        value="minimumStdevByTraceCount"
      >
        Remove Outliers
      </option>
    </select>
    <div
      class="optionsPanel"
    >
      <label
        for="startFrame"
      >
        Start frame
      </label>
      <input
        id="startFrame"
        max="5"
        min="1"
        type="number"
        value="1"
      />
      <label
        for="endFrame"
      >
        End frame
      </label>
      <input
        id="endFrame"
        max="5"
        min="1"
        type="number"
        value="3"
      />
      <label
        for="percentChange"
      >
        % of total change
      </label>
      <input
        id="percentChange"
        max="100"
        min="1"
        step="1"
        type="number"
        value="10"
      />
    </div>
  </div>
</div>
`;

exports[`component SelectionView render selection choices - first channel 3`] = `
<div>
  <div
    id="selectionPanel"
  >
    <select
      id="selectionMode"
    >
      <option
        value="manual"
      >
        Manual
      </option>
      <option
        value="percentChange"
      >
        % Change
      </option>
      <option
        value="stdev"
      >
        STDEV
      </option>
      <option
        value="minimumStdevByTraceCount"
      >
        Remove Outliers
      </option>
    </select>
    <div
      class="optionsPanel"
    >
      <label
        for="startBaselineFrame"
      >
        Start baseline frame
      </label>
      <input
        id="startBaselineFrame"
        max="5"
        min="1"
        type="number"
        value="1"
      />
      <label
        for="endBaselineFrame"
      >
        End baseline frame
      </label>
      <input
        id="endBaselineFrame"
        max="5"
        min="1"
        type="number"
        value="3"
      />
      <label
        for="startDetectionFrame"
      >
        Start detection frame
      </label>
      <input
        id="startDetectionFrame"
        max="5"
        min="1"
        type="number"
        value="4"
      />
      <label
        for="endDetectionFrame"
      >
        End detection frame
      </label>
      <input
        id="endDetectionFrame"
        max="5"
        min="1"
        type="number"
        value="5"
      />
      <label
        for="stdevMultiple"
      >
        x STDEV minimum
      </label>
      <input
        id="stdevMultiple"
        max="10"
        min="0.1"
        step="0.1"
        type="number"
        value="1"
      />
    </div>
  </div>
</div>
`;

exports[`component SelectionView render selection choices - first channel 4`] = `
<div>
  <div
    id="selectionPanel"
  >
    <select
      id="selectionMode"
    >
      <option
        value="manual"
      >
        Manual
      </option>
      <option
        value="percentChange"
      >
        % Change
      </option>
      <option
        value="stdev"
      >
        STDEV
      </option>
      <option
        value="minimumStdevByTraceCount"
      >
        Remove Outliers
      </option>
    </select>
    <div
      class="optionsPanel"
    >
      <label
        for="selectedTraceCount"
      >
        Selected trace count
      </label>
      <input
        id="selectedTraceCount"
        max="4"
        min="2"
        step="1"
        type="number"
        value="4"
      />
      <label
        id="selectedStdevLabel"
      >
        STDEV of selected
      </label>
      <div
        aria-labelledby="selectedStdevLabel"
        id="selectedStdev"
      >
        2.49
      </div>
    </div>
  </div>
</div>
`;

exports[`component SelectionView render selection choices - second channel 1`] = `
<div>
  <div
    id="selectionPanel"
  >
    <select
      id="selectionMode"
    >
      <option
        value="manual"
      >
        Manual
      </option>
      <option
        value="percentChange"
      >
        % Change
      </option>
      <option
        value="stdev"
      >
        STDEV
      </option>
      <option
        value="minimumStdevByTraceCount"
      >
        Remove Outliers
      </option>
    </select>
    <div
      class="optionsPanel"
    />
  </div>
</div>
`;

exports[`component SelectionView render selection choices - second channel 2`] = `
<div>
  <div
    id="selectionPanel"
  >
    <select
      id="selectionMode"
    >
      <option
        value="manual"
      >
        Manual
      </option>
      <option
        value="percentChange"
      >
        % Change
      </option>
      <option
        value="stdev"
      >
        STDEV
      </option>
      <option
        value="minimumStdevByTraceCount"
      >
        Remove Outliers
      </option>
    </select>
    <div
      class="optionsPanel"
    >
      <label
        for="startFrame"
      >
        Start frame
      </label>
      <input
        id="startFrame"
        max="5"
        min="1"
        type="number"
        value="1"
      />
      <label
        for="endFrame"
      >
        End frame
      </label>
      <input
        id="endFrame"
        max="5"
        min="1"
        type="number"
        value="3"
      />
      <label
        for="percentChange"
      >
        % of total change
      </label>
      <input
        id="percentChange"
        max="100"
        min="1"
        step="1"
        type="number"
        value="10"
      />
    </div>
  </div>
</div>
`;

exports[`component SelectionView render selection choices - second channel 3`] = `
<div>
  <div
    id="selectionPanel"
  >
    <select
      id="selectionMode"
    >
      <option
        value="manual"
      >
        Manual
      </option>
      <option
        value="percentChange"
      >
        % Change
      </option>
      <option
        value="stdev"
      >
        STDEV
      </option>
      <option
        value="minimumStdevByTraceCount"
      >
        Remove Outliers
      </option>
    </select>
    <div
      class="optionsPanel"
    >
      <label
        for="startBaselineFrame"
      >
        Start baseline frame
      </label>
      <input
        id="startBaselineFrame"
        max="5"
        min="1"
        type="number"
        value="1"
      />
      <label
        for="endBaselineFrame"
      >
        End baseline frame
      </label>
      <input
        id="endBaselineFrame"
        max="5"
        min="1"
        type="number"
        value="3"
      />
      <label
        for="startDetectionFrame"
      >
        Start detection frame
      </label>
      <input
        id="startDetectionFrame"
        max="5"
        min="1"
        type="number"
        value="4"
      />
      <label
        for="endDetectionFrame"
      >
        End detection frame
      </label>
      <input
        id="endDetectionFrame"
        max="5"
        min="1"
        type="number"
        value="5"
      />
      <label
        for="stdevMultiple"
      >
        x STDEV minimum
      </label>
      <input
        id="stdevMultiple"
        max="10"
        min="0.1"
        step="0.1"
        type="number"
        value="1"
      />
    </div>
  </div>
</div>
`;

exports[`component SelectionView render selection choices - second channel 4`] = `
<div>
  <div
    id="selectionPanel"
  >
    <select
      id="selectionMode"
    >
      <option
        value="manual"
      >
        Manual
      </option>
      <option
        value="percentChange"
      >
        % Change
      </option>
      <option
        value="stdev"
      >
        STDEV
      </option>
      <option
        value="minimumStdevByTraceCount"
      >
        Remove Outliers
      </option>
    </select>
    <div
      class="optionsPanel"
    >
      <label
        for="selectedTraceCount"
      >
        Selected trace count
      </label>
      <input
        id="selectedTraceCount"
        max="4"
        min="2"
        step="1"
        type="number"
        value="4"
      />
      <label
        id="selectedStdevLabel"
      >
        STDEV of selected
      </label>
      <div
        aria-labelledby="selectedStdevLabel"
        id="selectedStdev"
      >
        2.49
      </div>
    </div>
  </div>
</div>
`;
